File: autopilot.js
// 1
function getNewCar () {
  return {
    'City': 'Toronto',
    'Passengers': 0,
    'Gas': 100,
  }
}


// 2
function addCar(cars, newCar) {
  cars.push(newCar);
  console.log('Adding new car to fleet. Fleet size is now ' + cars.length + '.');
}


// 3
function pickUpPassenger(car) {
  car['Passengers'] += 1;
  car['Gas'] -= 10;
  console.log('Picked up passenger. Car now has ' + car['Passengers'] + ' passengers.');
}


// 4
function getDestination(car) {
  if (car['City'] === 'Toronto') {
    return 'Mississauga'
  } else if (car['City'] === 'Mississauga') {
    return 'London'
  } else if (car['City'] === 'London') {
    return 'Toronto'
  }
}


// 5

function fillUpGas(car) {
  var oldGas = car['Gas'];
  car['Gas'] = 100;
  return 'Filled up to ' + getGasDisplay(car['Gas']) + ' on gas from ' + getGasDisplay(oldGas)
}

// 6
function getGasDisplay(gasAmount) {
  return gasAmount + '%'
}


// 7
function drive(car, cityDistance) {
  if (car['Gas'] < cityDistance) {
    return fillUpGas(car);
  }

  car['City'] = getDestination(car);
  car['Gas'] -= cityDistance;
  return 'Drove to ' + car['City'] + '. Remaining gas: ' + getGasDisplay(car['Gas'])
}


// 8
function dropOffPassengers(car) {
  var previousPassengers = car['Passengers'];
  car['Passengers'] = 0;
  return 'Dropped off ' + previousPassengers + ' passengers.'
}


// 9
function act(car) {
  var distanceBetweenCities = 50;

  if (car['Gas'] < 20) {
    fillUpGas(car);
  } else if (car['Passengers'] < 3) {
    pickUpPassenger(car);
  } else {
    if (car['Gas'] < distanceBetweenCities) {
      return fillUpGas(car);
    }
    var droveTo = drive(car, distanceBetweenCities);
    var passengersDropped = dropOffPassengers(car);
    return droveTo + passengersDropped;
  }
}

// 10
function commandFleet(cars){
  for (i = 0; i < cars.length; i++){
    var car = cars[i];
    var action = act(car);
     console.log('Car ' + (i + 1) + ': ' + action )
  }
  console.log('---')
}


// 11
function addOneCarPerDay(cars, num_days) {
  for (var i = 0; i < num_days; i++) {
    var new_car = getNewCar();
    console.log(addCar(cars, new_car));
    commandFleet(cars);
  }
}



var cars = [];
addOneCarPerDay(cars, 10)

File: autopilot.rb
# 1
def get_new_car
  {
    city: 'Toronto',
    passengers: 0,
    gas: 100,
  }
end


# 2
def add_car(cars, new_car)
  cars << new_car
  "Adding new car to fleet. Fleet size is now #{cars.length}."
end


# 3
def pick_up_passenger(car)
  car[:passengers] += 1
  car[:gas] -= 10
  "Picked up passenger. Car now has #{car[:passengers]} passengers."
end


# 4
def get_destination(car)
  if car[:city] == 'Toronto'
    'Mississauga'
  elsif car[:city] == 'Mississauga'
    'London'
  elsif car[:city] == 'London'
    'Toronto'
  end
end


# 5
def fill_up_gas(car)
  old_gas = car[:gas]
  car[:gas] = 100
  "Filled up to #{ get_gas_display(car[:gas]) } on gas from #{ get_gas_display(old_gas) }."
end


# 6
def get_gas_display(gas_amount)
  "#{gas_amount}%"
end


# 7
def drive(car, city_distance)
  if car[:gas] < city_distance
    return fill_up_gas(car)
  end

  car[:city] = get_destination(car)
  car[:gas] -= city_distance
  "Drove to #{car[:city]}. Remaining gas: #{ get_gas_display(car[:gas]) }."
end


# 8
def drop_off_passengers(car)
  previous_passengers = car[:passengers]
  car[:passengers] = 0
  "Dropped off #{previous_passengers} passengers."
end


# 9
def act(car)
  distance_between_cities = 50

  if car[:gas] < 20
    fill_up_gas(car)

  elsif car[:passengers] < 3
    pick_up_passenger(car)

  else
    if car[:gas] < distance_between_cities
      return fill_up_gas(car)
    end
    drove_to = drive(car, distance_between_cities)
    passengers_dropped = drop_off_passengers(car)
    "#{drove_to} #{passengers_dropped}"
  end
end


# 10
def command_fleet(cars)
  cars.each_with_index do |car, i|
    action = act(car)
    puts "Car #{i + 1}: #{action}"
  end
  puts '---'
end


# 11
def add_one_car_per_day(cars, num_days)
  num_days.times do
    new_car = get_new_car
    puts add_car(cars, new_car)
    command_fleet(cars)
  end
end

cars = []
add_one_car_per_day(cars, 10)
